src/drivers/distance_sensor/cm8jl65/cm8jl65.cpp

```diff
@@ -0,0 +1,784 @@
+						return -EINVAL;
+					}
+
+					start();
+
+
+					return OK;
+				}
+			}
+		}
+
+	default:
+		/* give it to the superclass */
+		return CDev::ioctl(filp, cmd, arg);
+	}
+}
+
+int
+
+/*
+ * Method: collect()
+ *
+ * This method reads data from serial UART and places it into a buffer
+*/
+CM8JL65::collect()
+{
+	int bytes_read = 0;
+	int bytes_processed = 0;
+	int i = 0;
+	bool crc_valid = false;
+
+
+	perf_begin(_sample_perf);
+
+
+	/* read from the sensor (uart buffer) */
+	bytes_read = ::read(_fd, &_linebuf[0], sizeof(_linebuf));
+
+
+	if (bytes_read < 0) {
+		PX4_DEBUG("read err: %d \n", bytes_read);
+		perf_count(_comms_errors);
+		perf_end(_sample_perf);
+
+	} else if (bytes_read > 0) {
+		//     printf("Bytes read: %d \n",bytes_read);
+		i = bytes_read - 6 ;
+
+		while ((i >= 0) && (!crc_valid)) {
+			if (_linebuf[i] == START_FRAME_DIGIT1) {
+				bytes_processed = i;
+
+				while ((bytes_processed < bytes_read) && (!crc_valid)) {
+					//    printf("In the cycle, processing  byte %d, 0x%02X \n",bytes_processed, _linebuf[bytes_processed]);
+					if (OK == cm8jl65_parser(_linebuf[bytes_processed], _frame_data, &_parse_state, &_crc16, &_distance_mm)) {
+						crc_valid = true;
+					}
+
+					bytes_processed++;
+				}
+
+				_parse_state = STATE0_WAITING_FRAME;
+
+			}
+
+			// else {printf("Starting frame wrong. Index: %d value 0x%02X \n",i,_linebuf[i]);}
+			i--;
+		}
+
+	}
+
+	if (!crc_valid) {
+		return -EAGAIN;
+	}
+
+
+	//printf("val (int): %d, raw: 0x%08X, valid: %s \n", _distance_mm, _frame_data, ((crc_valid) ? "OK" : "NO"));
+
+	struct distance_sensor_s report;
+
+	report.timestamp = hrt_absolute_time();
+	report.type = distance_sensor_s::MAV_DISTANCE_SENSOR_LASER;
+	report.orientation = _rotation;
+	report.current_distance = _distance_mm / 1000.0f;
+	report.min_distance = get_minimum_distance();
+	report.max_distance = get_maximum_distance();
+	report.covariance = 0.0f;
+	report.signal_quality = -1;
+	/* TODO: set proper ID */
+	report.id = 0;
+
+	/* publish it */
+	orb_publish(ORB_ID(distance_sensor), _distance_sensor_topic, &report);
+
+	_reports->force(&report);
+
+	/* notify anyone waiting for data */
+	poll_notify(POLLIN);
+
+	bytes_read = OK;
+
+	perf_end(_sample_perf);
+
+	/* ENABLE THIS IF YOU WANT TO PRINT OLD VALUES WHILE CRC CHECK IS WRONG
+	if (!crc_valid) {
+			return -EAGAIN;
+		}
+	else return OK; */
+	return OK;
+
+}
+
+void
+CM8JL65::start()
+{
+	PX4_INFO("driver started");
+
+	_reports->flush();
+
+	/* schedule a cycle to start things */
+	work_queue(HPWORK, &_work, (worker_t)&CM8JL65::cycle_trampoline, this, 1);
+
+}
+
+void
+CM8JL65::stop()
+{
+	work_cancel(HPWORK, &_work);
+}
+
+void
+CM8JL65::cycle_trampoline(void *arg)
+{
+	CM8JL65 *dev = static_cast<CM8JL65 *>(arg);
+
+	dev->cycle();
+}
+
+void
+CM8JL65::cycle()
+{
+	//PX4_DEBUG("CM8JL65::cycle() - in the cycle");
+	/* fds initialized? */
+	if (_fd < 0) {
+		/* open fd */
+		_fd = ::open(_port, O_RDWR);
```

Comment:
This is not good. It means that the `read` call will block if there is no data available, and therefore will block the entire work queue.
Please change this to:
```
_fd = ::open(_port, O_RDWR | O_NOCTTY | O_NONBLOCK);
```

Reply:
Good to know, thanks!